flake8_trio/visitors/visitor91x.py

```diff
@@ -54,148 +106,232 @@ def __init__(self, *args: Any, **kwargs: Any):
+        else:
+            msg = node.__class__.__name__.lower()
+
+        self.error(
+            node,
+            msg,
+            statement,
+            error_code="TRIO911" if self.has_yield else "TRIO910",
+        )
+
+    def leave_Await(
+        self, original_node: cst.Await, updated_node: cst.Await
+    ) -> cst.Await:
         # the expression being awaited is not checkpointed
         # so only set checkpoint after the await node
-        self.generic_visit(node)
 
         # all nodes are now checkpointed
         self.uncheckpointed_statements = set()
+        return updated_node
 
     # raising exception means we don't need to checkpoint so we can treat it as one
-    visit_Raise = visit_Await
+    # can't use TypeVar due to libcst's built-in type checking not supporting it
+    leave_Raise = leave_Await  # type: ignore
 
     # Async context managers can reasonably checkpoint on either or both of entry and
     # exit.  Given that we can't tell which, we assume "both" to avoid raising a
     # missing-checkpoint warning when there might in fact be one (i.e. a false alarm).
-    def visit_AsyncWith(self, node: ast.AsyncWith):
-        self.visit_nodes(node.items)
-
-        self.uncheckpointed_statements = set()
-        self.visit_nodes(node.body)
+    def visit_With_body(self, node: cst.With):
+        if getattr(node, "asynchronous", None):
+            self.uncheckpointed_statements = set()
 
-        self.uncheckpointed_statements = set()
+    leave_With_body = visit_With_body
 
     # error if no checkpoint since earlier yield or function entry
-    def visit_Yield(self, node: ast.Yield):
+    def leave_Yield(
+        self, original_node: cst.Yield, updated_node: cst.Yield
+    ) -> cst.Yield:
         if not self.async_function:
-            return
+            return updated_node
         self.has_yield = True
-        self.generic_visit(node)
         for statement in self.uncheckpointed_statements:
-            self.error(
-                node,
-                "yield",
-                statement,
-                error_code="TRIO911",
-            )
+            self.error_91x(original_node, statement)
 
         # mark as requiring checkpoint after
-        self.uncheckpointed_statements = {
-            Statement("yield", node.lineno, node.col_offset)
-        }
+        pos = self.get_metadata(PositionProvider, original_node).start
+        self.uncheckpointed_statements = {Statement("yield", pos.line, pos.column)}
+        return updated_node
 
     # valid checkpoint if there's valid checkpoints (or raise) in:
     # (try or else) and all excepts, or in finally
     #
     # try can jump into any except or into the finally* at any point during it's
     # execution so we need to make sure except & finally can handle worst-case
     # * unless there's a bare except / except BaseException - not implemented.
-    def visit_Try(self, node: ast.Try):
+    def visit_Try(self, node: cst.Try):
```

Comment:
visit_Try was previously a massive function, it's now split into several different attribute visitor and leavers - but retains pretty much the exact same logic.